File: BOJ/Gold IV/BOJ2580.py
#BOJ2580 스도쿠 20210323 pypy3로 제출함
import sys
input = sys.stdin.readline

li = []

def isValid(r,c,v):
    for i in range((r//3)*3, (r//3)*3+3):
        for j in range((c//3)*3, (c//3)*3+3):
            if li[i][j] == v: return False
    for i in range(9):
        if li[r][i] == v: return False
        if li[i][c] == v: return False
    return True

def dfs(blank):
    global li
    if not blank:
        for v in li:
            print(*v)
        exit()
    r,c,can = blank[0]
    for v in can:
        if isValid(r,c,v):
            li[r][c] = v
            dfs(blank[1:])
            li[r][c] = 0


def main():
    global li
    li = [list(map(int, input().rstrip().split())) for _ in range(9)]
    blank = []
    for i in range(9):
        for j in range(9):
            if li[i][j] == 0:
                tmp = set()
                for ii in range((i//3)*3, (i//3)*3+3):
                    for jj in range((j//3)*3, (j//3)*3+3):
                        tmp.add(li[ii][jj])
                for ii in range(9):
                    tmp.add(li[i][ii])
                    tmp.add(li[ii][j])
                can = []
                for v in range(1,10):
                    if v not in tmp:
                        can.append(v)
                blank.append((i,j,can))
    dfs(blank)
if __name__ == '__main__':
    main()

File: BOJ/Gold IV/BOJ2580.java
//BOJ2580 스도쿠 20210706
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

import static java.lang.Math.max;

public class Main {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    private static int[][] board = new int[9][9];
    private static List<int[]> zeroes;
    private static Map<int[], List<Integer>> cases;

    public static void main(String[] args) throws IOException {
        for(int i = 0; i < 9; i++){
            board[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }

        zeroes = getZeroes();
        cases = getCases(zeroes);

        dfs(0, zeroes.size());

        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 8; j++){
                bw.write(board[i][j] + " ");
            }
            bw.write(board[i][8] + "\n");
        }

        bw.flush();
        bw.close();
    }

    public static boolean dfs(int v, int end){
        if(v == end) return true;
        int[] point = zeroes.get(v);
        int r = point[0], c = point[1];
        for(int can: cases.get(point)){
            if(isValid(r,c,can)){
                board[r][c] = can;
                if(dfs(v+1, end)) return true;
                board[r][c] = 0;
            }
        }
        return false;
    }

    public static boolean isValid(int r, int c, int v){
        for(int i = 0; i < 9; i++){
            if(board[i][c] == v) return false;
            if(board[r][i] == v) return false;
        }
        int sr = (r/3) * 3;
        int sc = (c/3) * 3;
        for(int i = sr; i < sr + 3; i++){
            for(int j = sc; j < sc + 3; j++){
                if(board[i][j] == v) return false;
            }
        }
        return true;
    }

    public static List<int[]> getZeroes(){
        List<int[]> zeroes = new ArrayList<>();
        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] == 0){
                    zeroes.add(new int[]{i,j});
                }
            }
        }
        return zeroes;
    }

    public static Map<int[], List<Integer>> getCases(List<int[]> zeroes){
        Map<int[], List<Integer>> ret = new HashMap<>();

        for(int[] point: zeroes){
            ret.put(point, new ArrayList<>());
            Set<Integer> cant = new HashSet<>();
            int r = point[0], c = point[1];
            for(int i = 0; i < 9; i++){
                cant.add(board[i][c]);
                cant.add(board[r][i]);
            }
            int sr = (r/3) * 3;
            int sc = (c/3) * 3;
            for(int i = sr; i < sr + 3; i++){
                for(int j = sc; j < sc + 3; j++){
                    cant.add(board[i][j]);
                }
            }
            for(int i = 1; i <= 9; i++){
                if(!cant.contains(i)) ret.get(point).add(i);
            }
        }

        return ret;
    }
}